sno/structs.py

```diff
@@ -72,9 +72,55 @@ def resolve_refish(repo, refish):
         return CommitWithReference(repo.resolve_refish(refish))
 
 
-# When merging two commits, we have three versions of lots of things - these three versions
-# are the common ancestor, and the two versions to be merged, "ours" and "theirs".
-AncestorOursTheirs = namedtuple("AncestorOursTheirs", ("ancestor", "ours", "theirs"))
+# pygit2 always has this order - we use it too for consistency,
+# and so we can meaningfully zip() our tuples with theirs
+_ANCESTOR_OURS_THEIRS_ORDER = ("ancestor", "ours", "theirs")
 
-AncestorOursTheirs.names = AncestorOursTheirs._fields
-AncestorOursTheirs.chars = tuple(n[0] for n in AncestorOursTheirs.names)
+
+class AncestorOursTheirs(namedtuple("AncestorOursTheirs", _ANCESTOR_OURS_THEIRS_ORDER)):
+    """
+    When merging two commits, we can end up with three versions of lots of things -
+    commits, repository-structures, datasets, features, primary keys.
+    The 3 versions  are the common ancestor, and 2 versions to be merged, "ours" and "theirs".
+    Like pygit2, we keep the 3 versions always in the same order - ancestor, ours, theirs.
+    """
+
+    NAMES = _ANCESTOR_OURS_THEIRS_ORDER
+    CHARS = tuple(n[0] for n in NAMES)
+
+    @staticmethod
+    def of(*args, **kwargs):
+        """
+        Coerces any of the following into an AncestorOursTheirs:
+        3 positional args, 3 keyword args, a tuple, a dict, or a generator.
```

Comment:
Seems unnecessarily loose, confuses just exactly what this thing accepts (it accepts everything!). These could all just fold back to "use either *args or **kwargs" I think.

generators can be starred into args by the caller: `f(*gen())`
dicts can be double-starred by the caller: `f(**mydict)`

If caller can do that then I don't really know what this method adds over the existing namedtuple `__init__`

Reply:
Removed.

But, I did instead make a little function in conflicts.py which lets you use generators easily, so I can do this:
```
datasets3 = aot(
    _safe_get_dataset_for_index_entry(rs, ie)
    for rs, ie in zip(repo_structures3, index_entries3)
)
```
instead of this:
```
datasets3 = AncestorOursTheirs(
    *(
        _safe_get_dataset_for_index_entry(rs, ie)
        for rs, ie in zip(repo_structures3, index_entries3)
    )
)
```